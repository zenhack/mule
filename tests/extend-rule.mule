#XFAIL: type formatting
let
  # These two types should be equivalent, and both assignable
  # to the given expression:
  , a : { y : int } -> { y : int } = fn x. x where { y = 2 }
  , b : {         } -> { y : int } = fn x. x where { y = 2 }
in
{}
