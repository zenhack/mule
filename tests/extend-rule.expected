
Desugared:

	(letrec (types ())
	 (values
	  ((b
	    ((:
	      ((record (types ()) (values ())) ->
	       (record (types ()) (values ((y int))))))
	     (fn x (((.= y) x) (Integer 2)))))
	   (a
	    ((:
	      ((record (types ()) (values ((y int)))) ->
	       (record (types ()) (values ((y int))))))
	     (fn x (((.= y) x) (Integer 2)))))))
	 {})

inferred type:

	(all a (record (types (...a)) (values ())))

Runtime term:

	(LetRec
	 (((0
	    (Lam
	     (0 ()
	      (App
	       ((App
	         ((Lam
	           (0 ()
	            (Lam (1 () (Update (old (Var 1)) (label y) (field (Var 0)))))))
	          (Var 0)))
	        (Const (Integer 2)))))))
	   (0
	    (Lam
	     (0 ()
	      (App
	       ((App
	         ((Lam
	           (0 ()
	            (Lam (1 () (Update (old (Var 1)) (label y) (field (Var 0)))))))
	          (Var 0)))
	        (Const (Integer 2))))))))
	  (Record ())))

Evaluated:

	(Record ())
